pkg/util/metrics.go

```diff
@@ -0,0 +1,187 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"net/http"
+	"strings"
+	"sync"
+
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	prometheus_model "github.com/prometheus/client_model/go"
+	"k8s.io/client-go/util/workqueue"
+)
+
+func CreateValidMetricNameLabel(prefix, name string) string {
+	// "-" aren't valid characters for prometheus metric names or labels
+	return strings.Replace(prefix+name, "-", "_", -1)
+}
+
+// Best effort metric registration with Prometheus.
+func RegisterMetric(metric prometheus.Collector) {
+	if err := prometheus.Register(metric); err != nil {
+		glog.Errorf("Error while registering Prometheus metric: [%v]", err)
+	}
+}
+
+type MetricConfig struct {
+	MetricsEndpoint string
+	MetricsPort     string
+	MetricsPrefix   string
+	MetricsLabels   []string
+}
+
+type PositiveGauge struct {
+	mux         sync.RWMutex
+	name        string
+	gaugeMetric *prometheus.GaugeVec
+}
+
+// Gauge with conditional decrement ensuring its value is never negative.
+func NewPositiveGauge(name string, description string, labels []string) *PositiveGauge {
+	for i, label := range labels {
+		labels[i] = CreateValidMetricNameLabel("", label)
+	}
+
+	gauge := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: name,
+			Help: description,
+		},
+		labels,
+	)
+
+	RegisterMetric(gauge)
+
+	return &PositiveGauge{
+		gaugeMetric: gauge,
+		name:        name,
+	}
+}
+
+func fetchGaugeValue(m *prometheus.GaugeVec, labels map[string]string) float64 {
+	// Hack to get the current value of the metric to support PositiveGauge
+	pb := &prometheus_model.Metric{}
+
+	m.With(labels).Write(pb)
+	return pb.GetGauge().GetValue()
+}
+
+func (c *PositiveGauge) Value(labelMap map[string]string) float64 {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return fetchGaugeValue(c.gaugeMetric, labelMap)
+}
+
+// Increment the Metric for the labels specified
+func (c *PositiveGauge) Inc(labelMap map[string]string) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	if m, err := c.gaugeMetric.GetMetricWith(labelMap); err != nil {
+		glog.Errorf("Error while exporting metrics: %v", err)
+	} else {
+		glog.V(2).Infof("Incrementing %s with labels %s", c.name, labelMap)
+		m.Inc()
+	}
+}
+
+// Decrement the metric only if its positive for the labels specified
+func (c *PositiveGauge) Dec(labelMap map[string]string) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	// Decrement only if positive
+	val := fetchGaugeValue(c.gaugeMetric, labelMap)
+	if val > 0 {
+		glog.V(2).Infof("Decrementing %s with labels %s metricVal to %v", c.name, labelMap, val-1)
```

Comment:
Looks like indention is not consistent: 2 in some places, whereas 4 in other places. Please run `go fmt ./...` to format the code.

Reply:
done